src/Battlescape/SkillMenuState.cpp

```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright 2010-2020 OpenXcom Developers.
+ *
+ * This file is part of OpenXcom.
+ *
+ * OpenXcom is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * OpenXcom is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with OpenXcom.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "SkillMenuState.h"
+#include "../Engine/Game.h"
+#include "../Engine/Options.h"
+#include "../Engine/LocalizedText.h"
+#include "../Engine/Action.h"
+#include "../Engine/Sound.h"
+#include "../Engine/Unicode.h"
+#include "../Savegame/BattleUnit.h"
+#include "../Savegame/BattleItem.h"
+#include "../Savegame/SoldierDiary.h"
+#include "../Mod/Mod.h"
+#include "../Mod/RuleSkill.h"
+#include "../Mod/RuleSoldier.h"
+#include "../Mod/RuleSoldierBonus.h"
+#include "ActionMenuItem.h"
+#include "PrimeGrenadeState.h"
+#include "MedikitState.h"
+#include "ScannerState.h"
+#include "../Savegame/SavedGame.h"
+#include "../Savegame/SavedBattleGame.h"
+#include "../Savegame/Tile.h"
+#include "Pathfinding.h"
+#include "TileEngine.h"
+#include "../Interface/Text.h"
+#include "../Savegame/EquipmentLayoutItem.h"
+
+namespace OpenXcom
+{
+	
+/**
+ * Initializes all the elements in the Action Menu window.
+ * @param game Pointer to the core game.
+ * @param action Pointer to the action.
+ * @param x Position on the x-axis.
+ * @param y position on the y-axis.
+ */
+SkillMenuState::SkillMenuState(BattleAction *action, int x, int y) : ActionMenuState(action)
+{
+	_screen = false;
+	
+	// Set palette
+	_game->getSavedGame()->getSavedBattle()->setPaletteByDepth(this);
+
+	for (int i = 0; i < 6; ++i)
+	{
+		_actionMenu[i] = new ActionMenuItem(i, _game, x, y);
+		add(_actionMenu[i]);
+		_actionMenu[i]->setVisible(false);
+		_actionMenu[i]->onMouseClick((ActionHandler)&SkillMenuState::btnActionMenuItemClick);
+	}
```

Comment:
Two things, from where you get `< 6` value? Better would be use actual size of this array.
Second, isn't `_actionMenu` already filled by `ActionMenuState` constructor?

Reply:
From the code which is in the constructor of ActionMenuState. The code there uses 6. I just copied it over, assuming whoever wrote the code had a good reason to limit the menu to this amount.

No, since I call a new empty constructor of ActionMenuState which does not do this work.